File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Sudoku/Sudoku_CSharp/SolverGenerator/Techniques/BeginnerTechnique.cs
//--------------------------------------------------------------------------
// 
//  Copyright (c) Microsoft Corporation.  All rights reserved. 
// 
//  File: BeginnerTechnique.cs
//
//  Description: Implements the Sudoku beginner technique.
// 
//--------------------------------------------------------------------------

using System;
using System.Drawing;
using Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Collections;

namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Techniques
{
	/// <summary>Implements the beginner elimination technique.</summary>
	[Serializable]
    public sealed class BeginnerTechnique : EliminationTechnique
	{
		/// <summary>Initialize the technique.</summary>
		public BeginnerTechnique(){}

		/// <summary>Gets the difficulty level of this technique.</summary>
		internal override uint DifficultyLevel { get { return 1; } }

        public override bool Equals(object obj)
        {
            return obj is BeginnerTechnique;
        }

        public override int GetHashCode()
        {
            return GetType().GetHashCode();
        }

		/// <summary>Runs this elimination technique over the supplied puzzle state and previously computed possible numbers.</summary>
		/// <param name="state">The puzzle state.</param>
		/// <param name="exitEarlyWhenSoleFound">Whether the method can exit early when a cell with only one possible number is found.</param>
		/// <param name="possibleNumbers">The previously computed possible numbers.</param>
		/// <param name="numberOfChanges">The number of changes made by this elimination technique.</param>
		/// <param name="exitedEarly">Whether the method exited early due to a cell with only one value being found.</param>
		/// <returns>Whether more changes may be possible based on changes made during this execution.</returns>
		internal override bool Execute(
			PuzzleState state, bool exitEarlyWhenSoleFound,
			FastBitArray[][] possibleNumbers, out int numberOfChanges, out bool exitedEarly)
		{
			numberOfChanges = 0;
			exitedEarly = false;

			int numLocations;
			Point [] locations = new Point[state.GridSize];

			for(int n=0; n<state.GridSize; n++)
			{
				// Find all occurrences of n.  If GridSize or more are found, no point in continuing
				// on.  Obviously if more than GridSize are found, there's a mistake somewhere on the
				// board (at least one), but that's not our problem.
				numLocations = 0;
				for(int row=0; row<state.GridSize && numLocations < state.GridSize; row++)
				{
					for(int column=0; column<state.GridSize && numLocations < state.GridSize; column++)
					{
						if (state[row,column].HasValue &&
							state[row,column].Value == n) locations[numLocations++] = new Point(row, column);
					}
				}
				if (numLocations >= state.GridSize) continue;

				// For each box
				for(int box=0; box<state.GridSize; box++)
				{
					bool done = false;

					// If any of the cells in the box is the number, bail
					int boxStartX = (box % state.BoxSize) * state.BoxSize;
					for (int x = boxStartX; x < boxStartX + state.BoxSize && !done; x++)
					{
						int boxStartY = (box / state.BoxSize) * state.BoxSize;
						for (int y = boxStartY; y < boxStartY + state.BoxSize && !done; y++)
						{
							Point cell = new Point(x,y);
							if ((state[cell].HasValue && state[cell].Value == n) ||
								(!state[cell].HasValue && possibleNumbers[x][y].CountSet == 1 && 
								possibleNumbers[x][y][n])) done = true;
						}
					}
					if (done) continue;

					// Look at each cell in the box
					Point targetCell = new Point(-1,-1);
					boxStartX = (box % state.BoxSize) * state.BoxSize;
					for (int x = boxStartX; x < boxStartX + state.BoxSize && !done; x++)
					{
						int boxStartY = (box / state.BoxSize) * state.BoxSize;
						for (int y = boxStartY; y < boxStartY + state.BoxSize && !done; y++)
						{
                            // See if there's only one cell in the box that can be
                            // this number.  If there is, it must be there.
							Point cell = new Point(x,y);
							if (!state[cell].HasValue && possibleNumbers[x][y][n])
							{
								bool invalid = false;
								for(int locNum=0; locNum<numLocations; locNum++)
								{
									if (locations[locNum].X == cell.X || locations[locNum].Y == cell.Y) 
									{ 
										invalid = true;
										break;
									}
								}
								if (!invalid)
								{
									if (targetCell == new Point(-1,-1)) targetCell = cell;
									else
									{
										targetCell = new Point(-1,-1);
										done = true;
									}
								}
							}
						}
					}

                    // If we found a cell we can fill in, modify the possible numbers list appropriately
					if (targetCell != new Point(-1,-1))
					{
						possibleNumbers[targetCell.X][targetCell.Y].SetAll(false);
						possibleNumbers[targetCell.X][targetCell.Y][n] = true;
						numberOfChanges++;
						if (exitEarlyWhenSoleFound)
						{
							exitedEarly = true;
							return false;
						}
					}
				}
			}

            // If numberOfChanges > 0, we made a change to possibleNumbers, but the state
            // itself wasn't affected.  Thus, running this again won't help.
			return false;
		}
	}
}

File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Sudoku/Sudoku_VisualBasic/SolverGenerator/Techniques/BeginnerTechnique.vb
'--------------------------------------------------------------------------
' 
'  Copyright (c) Microsoft Corporation.  All rights reserved. 
' 
'  File: BeginnerTechnique.vb
'
'  Description: Implements the Sudoku beginner technique.
' 
'--------------------------------------------------------------------------

Imports Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Collections

Namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Techniques
	''' <summary>Implements the beginner elimination technique.</summary>
	<Serializable>
	Public NotInheritable Class BeginnerTechnique
		Inherits EliminationTechnique
		''' <summary>Initialize the technique.</summary>
		Public Sub New()
		End Sub

		''' <summary>Gets the difficulty level of this technique.</summary>
		Friend Overrides ReadOnly Property DifficultyLevel() As UInteger
			Get
				Return 1
			End Get
		End Property

		Public Overrides Overloads Function Equals(ByVal obj As Object) As Boolean
			Return TypeOf obj Is BeginnerTechnique
		End Function

		Public Overrides Function GetHashCode() As Integer
			Return Me.GetType().GetHashCode()
		End Function

		''' <summary>Runs this elimination technique over the supplied puzzle state and previously computed possible numbers.</summary>
		''' <param name="state">The puzzle state.</param>
		''' <param name="exitEarlyWhenSoleFound">Whether the method can exit early when a cell with only one possible number is found.</param>
		''' <param name="possibleNumbers">The previously computed possible numbers.</param>
		''' <param name="numberOfChanges">The number of changes made by this elimination technique.</param>
		''' <param name="exitedEarly">Whether the method exited early due to a cell with only one value being found.</param>
		''' <returns>Whether more changes may be possible based on changes made during this execution.</returns>
        Friend Overrides Function Execute(ByVal state As PuzzleState, ByVal exitEarlyWhenSoleFound As Boolean,
            ByVal possibleNumbers()() As FastBitArray, <System.Runtime.InteropServices.Out()> ByRef numberOfChanges As Integer,
            <System.Runtime.InteropServices.Out()> ByRef exitedEarly As Boolean) As Boolean
            numberOfChanges = 0
            exitedEarly = False

            Dim numLocations As Integer
            Dim locations(state.GridSize - 1) As Point

            For n = 0 To state.GridSize - 1
                ' Find all occurrences of n.  If GridSize or more are found, no point in continuing
                ' on.  Obviously if more than GridSize are found, there's a mistake somewhere on the
                ' board (at least one), but that's not our problem.
                numLocations = 0
                Dim row = 0
                Do While row < state.GridSize AndAlso numLocations < state.GridSize
                    Dim column = 0
                    Do While column < state.GridSize AndAlso numLocations < state.GridSize
                        If state(row, column).HasValue AndAlso state(row, column).Value = n Then
                            locations(numLocations) = New Point(row, column)
                            numLocations += 1
                        End If

                        column += 1
                    Loop
                    row += 1
                Loop
                If numLocations >= state.GridSize Then
                    Continue For
                End If

                ' For each box
                For box = 0 To state.GridSize - 1
                    Dim done = False

                    ' If any of the cells in the box is the number, bail
                    Dim boxStartX = (box Mod state.BoxSize) * state.BoxSize
                    Dim x = boxStartX
                    Do While x < boxStartX + state.BoxSize AndAlso Not done
                        Dim boxStartY = (box \ state.BoxSize) * state.BoxSize
                        Dim y = boxStartY
                        Do While y < boxStartY + state.BoxSize AndAlso Not done
                            Dim cell As New Point(x, y)
                            If (state(cell).HasValue AndAlso state(cell).Value = n) OrElse ((Not state(cell).HasValue) AndAlso
                                                possibleNumbers(x)(y).CountSet = 1 AndAlso possibleNumbers(x)(y)(n)) Then
                                done = True
                            End If
                            y += 1
                        Loop
                        x += 1
                    Loop
                    If done Then
                        Continue For
                    End If

                    ' Look at each cell in the box
                    Dim targetCell As New Point(-1, -1)
                    boxStartX = (box Mod state.BoxSize) * state.BoxSize
                    x = boxStartX
                    Do While x < boxStartX + state.BoxSize AndAlso Not done
                        Dim boxStartY = (box \ state.BoxSize) * state.BoxSize
                        Dim y = boxStartY
                        Do While y < boxStartY + state.BoxSize AndAlso Not done
                            ' See if there's only one cell in the box that can be
                            ' this number.  If there is, it must be there.
                            Dim cell As New Point(x, y)
                            If (Not state(cell).HasValue) AndAlso possibleNumbers(x)(y)(n) Then
                                Dim invalid = False
                                For locNum = 0 To numLocations - 1
                                    If locations(locNum).X = cell.X OrElse locations(locNum).Y = cell.Y Then
                                        invalid = True
                                        Exit For
                                    End If
                                Next locNum
                                If Not invalid Then
                                    If targetCell = New Point(-1, -1) Then
                                        targetCell = cell
                                    Else
                                        targetCell = New Point(-1, -1)
                                        done = True
                                    End If
                                End If
                            End If
                            y += 1
                        Loop
                        x += 1
                    Loop

                    ' If we found a cell we can fill in, modify the possible numbers list appropriately
                    If targetCell <> New Point(-1, -1) Then
                        possibleNumbers(targetCell.X)(targetCell.Y).SetAll(False)
                        possibleNumbers(targetCell.X)(targetCell.Y)(n) = True
                        numberOfChanges += 1
                        If exitEarlyWhenSoleFound Then
                            exitedEarly = True
                            Return False
                        End If
                    End If
                Next box
            Next n

            ' If numberOfChanges > 0, we made a change to possibleNumbers, but the state
            ' itself wasn't affected.  Thus, running this again won't help.
            Return False
        End Function
	End Class
End Namespace